File: 장세훈/LeetCode/easy/ValidParentheses.java
/*
        20. Valid Parentheses
        https://leetcode.com/problems/valid-parentheses/description/

        java는 Stack 대신 Dequq를 사용할 것을 권장.
        Stack은 모든 기능에 lock이 걸려 있어 성능상 안좋음
 */
class ValidParentheses {

    public static void main(String[] args) {
        new ValidParentheses().isValid("()");
    }
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        Map<Character, Character> map = new HashMap<>() {{
            put(')', '(');
            put(']', '[');
            put('}', '{');
        }};

        for (char c : s.toCharArray()) {
            if(!map.containsKey(c)) {
                stack.push(c);
            }else if(stack.isEmpty() || stack.pop() != (map.get(c))) {
                return false;
            }
        }

        return stack.isEmpty();
    }
}

File: 정수빈/LeetCode/easy/valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses
# 난이도: easy
# 시간: 2분

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c in "{[(":
                stack.append(c)
            elif stack and c in "}" and stack[-1] == "{":
                stack.pop()
            elif stack and c in "]" and stack[-1] == "[":
                stack.pop()
            elif stack and c in ")" and stack[-1] == "(":
                stack.pop()
            else:
                return False

        if stack:
            return False
        else:
            return True                            
                